### ISSUE TITLE
Crop/trim jams utility function

### ISSUE THREAD

**Issue by justinsalamon** [marl/jams#135]
For [scaper](https://github.com/justinsalamon/scaper) I need to crop/trim some audio files and their corresponding JAMS annotations. I started implementing a `crop_jams` method in scaper, but after chatting with @bmcfee it transpired that this might be something useful to have as part of the core jams library, and regardless there are some methodological questions best run through this crowd.

Question 1: is this something we want to implement as part of the core jams lib?
(this is the second time I need to crop jams annotations for a project, so at least based on my own experience this would be something very useful to have).

Question 2: regardless of where this is implemented, I can see (at least) two different approaches to implementation:
1. Edit the annotation "in place" by directly manipulating the annotation.data DataFrame.
2. Create a new annotation, copy over all metadata, and selectively adjust/add events based on the desired crop range.

The former has the advantage of not having to worry about replicating the annotation perfectly since the data frame (and `.duration` object variable) is manipulated directly. The latter has the advantage of (potentially) being independent of the internal storage mechanism, though it requires care in ensuring all annotation information is copied over. Also if we wanted it to be completely decoupled from the dataframe we might want/need to implement some form of `get` function or iterator to be able to access (and maybe even remove?) the annotation observations via an API.

This is a little time-critical for me, so thoughts/discussion are most welcome. Worst case scenario I'll implement this in scaper first and port it into jams at a later point if we decide that's the way we want to go.

**Comment by bmcfee** [comment]
I've also run aground of slicing jams annotations in this way in a few places, so it wouldn't be terrible to consolidate efforts in one place.

Here are my thoughts:

1. We should, wherever possible, strive to keep data immutable.  If we're going to be modifying annotations like this, it should return a new annotation, not modify in-place.  This argues for taking option 2.
2. In general, I think we should support both trimming to a time slice, and padding on the front or back.
3. I think this would be best implemented by object methods that return new instances with all fields appropriately modified.  So you could say ``sub_jam = jam.slice(start, end)`` or ``jam_pad = jam.pad(start_pad, end_pad)`` and get coherent data out without modifying ``jam``.

I know this is time-critical for @justinsalamon, but it's a rather large undertaking that I don't want to hold up the (already absurdly delayed) next release.  Can we slot this for 0.3?  If @justinsalamon wants to implement in a PR, that's fine too.

**Comment by justinsalamon** [comment]
> 1. We should, wherever possible, strive to keep data immutable. If we're going to be modifying annotations like this, it should return a new annotation, not modify in-place. This argues for taking option 2.

Sure

> I think this would be best implemented by object methods that return new instances with all fields appropriately modified. So you could say `sub_jam = jam.slice(start, end)` or `jam_pad = jam.pad(start_pad, end_pad)` and get coherent data out without modifying jam.

Makes sense. For my own selfish reasons I'll start working on slice first though.

> I know this is time-critical for @justinsalamon, but it's a rather large undertaking that I don't want to hold up the (already absurdly delayed) next release. Can we slot this for 0.3? If @justinsalamon wants to implement in a PR, that's fine too.

For my immediate work that would be ok (since I can just work against the PR branch), but releasing scaper would depend on this functionality and I'm concerned that waiting for 0.3 might stall it a little too much?

I'm planning on bashing this out today/next few days, how about we see how far I get by the end of this weekend to determine whether we can squeeze it into the upcoming release?

**Comment by bmcfee** [comment]
Sure.  We could also make a minor point release for this feature before 0.3.

**Comment by justinsalamon** [comment]
> Sure. We could also make a minor point release for this feature before 0.3.

👍 guess I better get started then... :)

**Comment by justinsalamon** [comment]
Question: if we implement this method at the JAMS object level, it provides an easy way to trim all annotations contained in the jam with one call. On the flip side, since different annotations can span different time ranges, this means that we it doesn't really make sense to check/enforce that the specified trim start/end times are within the duration of every annotation (as they are likely to be within the range of some annotation and outside the range of others) - which means that basically this might actually end up "padding" some of the annotations, unless we only apply the trimming to those annotations for which the start/end times are valid and raise warnings for the others, but I'm not sure I like this option because it can (will) create a mess.

The alternative would be to implement `trim` as a method of `Annotation`, allowing the user to selectively trim only specific annotations. Then we could strictly require that the specified start/end times be within the annotation's duration, and could also implement another method at the JAMS object level that will iterate over all annotations and try to trim them (but will raise an error if any are outside the specified trim range).

Thoughts?

**Comment by bmcfee** [comment]
> On the flip side, since different annotations can span different time ranges, this means that it doesn't really make sense to check/enforce that the specified trim start/end times are within the duration of every annotation

Not sure I follow you there.

Here's how I see it:

1. Implement `Annotation.trim(start, end)`.  The logic is, I think, as follows:
    - `my_end = ann.start + ann.duration`
    - `ann.start = min(0, ann.start - start)`
    - `ann.duration = min(end, my_end) - ann.start`
    - And then trim all jamsframe events accordingly.
    - You might want to parameterize this by `duration` instead of `end`.  It might also be simpler to just handle the pad logic in here as well, in case `end > ann.duration - ann.start`.

2. Implement `trim` in all the other jams objects, particularly FileMetadata.

3. The top-level `trim` just recurses down the object hierarchy trimming things as needed to populate the new object.

**Comment by justinsalamon** [comment]
> Not sure I follow you there.

Let me try again: do we want to impose any conditions on being able to tim an annotation? For example, you can only trim an annotation if the specified start/end times are within the annotation's current start/end times? So for example, if an annotation starts at 5 and ends at 15, you can't for example call `ann.trim(3, 8)` because the new start time is outside the original range of the annotation.

Alternatively, we can just not impose any limitations, which means that `trim` can also be used to pad annotations (e.g. `ann.trim(0, 20)` would pad the annotation given above as an example).

> `ann.start = min(0, ann.start - start)`

I'm not sure I follow the logic here, wouldn't it just be, for `Annotation.trim(start, end)`:
```
ann.time = min(0, start)
ann.duration = end - start
```

The third option (perhaps what you were getting at?) would be to return the intersection of the original annotation range and the desired trim, so that calling `trim(3,8)` on an annotation that spans [5,15] would return an annotation spanning [5,8]. I'm not sure that's a good option though?

**Comment by justinsalamon** [comment]
In case it makes things clearer, here's what `Annotation.trim()` might look like if we don't impose any conditions on trimming (i.e. trimming can also be padding):

```python
    def trim(self, start_time, end_time):
        '''
        Trim the annotation and return as new Annotation object. Trimming will
        result in the new annotation only containing observations that occur
        between `start_time` and `end_time`. Observations that start before
        `start_time` but end after it will be trimmed such that they start at
        `start_time`, and similarly observations that start before `end_time`
        but end after it will be trimmed to end at `end_time`. The new
        duration of the annotation will be `end_time - start_time`. Note that
        depending on the values of `start_time` and `end_time` this function
        can also pad the annotation. This function also copies over all the
        annotation metadata from the original annotation, and adds a list of
        tuples to the annotation's sandbox keyed by `trim`, which documents
        the trim parameters: each trim operation adds a tuple to this list
        containing two elements: the `start_time` and `end_time`.

        Parameters
        ----------
        start_time : float
            The new start time for the annotation
        end_time
            The new end time for the annotation

        Returns
        -------
        ann_trimmed : jams.Annotation
            The trimmed annotation, return as a new jams.Annotation object.
        '''
        # Create new annotation with same namespace/metadata
        ann_trimmed = Annotation(
            self.namespace,
            data=None,
            annotation_metadata=self.annotation_metadata,
            sandbox=self.sandbox,
            time=start_time,
            duration=end_time - start_time)

        # Selectively add observations based on their start time / duration
        for idx, obs in self.data.iterrows():

            obs_start = obs['time'].total_seconds()
            obs_end = obs_start + obs['duration'].total_seconds()

            if obs_start < end_time and obs_end > start_time:
                new_start = max(obs_start, start_time)
                new_end = min(obs_end, end_time)
                new_duration = new_end - new_start
                ann_trimmed.append(time=new_start,
                                   duration=new_duration,
                                   value=obs['value'],
                                   confidence=obs['confidence'])

            if 'trim' not in ann_trimmed.sandbox.keys():
                ann_trimmed.sandbox.update(trim=[(start_time, end_time)])
            else:
                ann_trimmed.sandbox.trim.append((start_time, end_time))

        return ann_trimmed
```

**Comment by justinsalamon** [comment]
p.s.
> Implement trim in all the other jams objects, particularly FileMetadata.

Do we want this? You could reasonably want to trim an annotation without trimming the corresponding audio file, in which case the file metadata would remain unchanged. I would delegate this to the user, i.e. if they are also trimming the audio file, then they should make sure to update the file metadata. That's basically what will happen inside scaper.

**Comment by bmcfee** [comment]
> Do we want this? You could reasonably want to trim an annotation without trimming the corresponding audio file, in which case the file metadata would remain unchanged.

If that's what you want to do, then it should be done by calling trim on the annotation of interest and adding it back to the existing jams.

At the top level, trimming a jams should trim the entire object (including metadata) so that everything stays consistent.

**Comment by justinsalamon** [comment]
> At the top level, trimming a jams should trim the entire object (including metadata) so that everything stays consistent.

Was there ever any assumption that all annotations in a jams file have consistent time/duration? Isn't it the case that the `file_metadata` should always correspond to the audio file, but the annotations can span any time range? Under this scenario I think you'd only want to change the `file_metadata` if you actually manipulated the audio file - but I don't see why manipulating the annotations (specifically trimming/padding) should change the file metadata. E.g., you have a 3 min audio file and an annotation that spans the first 2 min, then for whatever reason you trim it to span just the first 1 min - the audio file is still 3 min. That's why I'd delegate modifying the `file_metadata` to the code that modifies the audio file (e.g. MUDA or Scaper).

But anyway, before we figure that one out the most pertinent decision to make is whether we want to impose limitations on trimming (so that it's only really trimming) or not (in which case it's trimming/padding).

Right now i've implemented trimming at the annotation level without limitations (so it's trimming/padding) as per the snippet I pasted in [this comment](https://github.com/marl/jams/issues/135#issuecomment-254929609).

**Comment by bmcfee** [comment]
> Was there ever any assumption that all annotations in a jams file have consistent time/duration?

Yes; at least, time within a single jams object should be consistent across all annotations.

It's okay for different annotations to span different subsets of the track: that's why we have `Annotation.time,duration` as well as within the observations.



> Isn't it the case that the file_metadata should always correspond to the audio file, but the annotations can span any time range?

This is syntactically allowed, but should not happen.



> Under this scenario I think you'd only want to change the file_metadata if you actually manipulated the audio file - but I don't see why manipulating the annotations (specifically trimming/padding) should change the file metadata. E.g., you have a 3 min audio file and an annotation that spans the first 2 min, then for whatever reason you trim it to span just the first 1 min - the audio file is still 3 min.

If you do this, then you should trim the annotation, not the jams object.

However, if you actually are trimming/padding a jams object to match an audio file (say, after doing some alignment / deskewing or what have you), then it's totally sensible to want to change the entire object at once.


> That's why I'd delegate modifying the file_metadata to the code that modifies the audio file (e.g. MUDA or Scaper).

I think it's a useful enough operation to live inside jams.



> But anyway, before we figure that one out the most pertinent decision to make is whether we want to impose limitations on trimming (so that it's only really trimming) or not (in which case it's trimming/padding).

I only mentioned padding in the previous post because it would simplify the logic a bit if you allow end to grow as well as shrink.

In general, padding is a separate problem, and should probably be a separate function.

**Comment by justinsalamon** [comment]
> I only mentioned padding in the previous post because it would simplify the logic a bit if you allow end to grow as well as shrink.
In general, padding is a separate problem, and should probably be a separate function.

So I already have this implemented (in a way that supports both trimming/padding) [here](https://github.com/marl/jams/blob/trim/jams/core.py#L872).

This could be changed to only support trimming (by adding checks to ensure `start_time` and `end_time` are within the time range spanned by the annotation): **let me know if that's how you want to go about it**.

By "consistent time" I didn't mean a consistent time-grid, rather a consistent *start time* (which in jams-land is just called "time" everywhere) and duration, to which, as you noted in your second reply, the answer is no (i.e. different annotations can span different time ranges).

> However, if you actually are trimming/padding a jams object to match an audio file (say, after doing some alignment / deskewing or what have you), then it's totally sensible to want to change the entire object at once.

ok, I'll go with this then.

**Comment by justinsalamon** [comment]
After discussing further with @bmcfee , we've agreed on the following functionality:

* `Annotation.trim()` (and hence `JAMS.trim()`) can only be used for trimming, not padding
* calling `ann.trim(start_time, end_time)` will return the *intersection* of the annotation's original time range (defined by `ann.time` and `ann.duration`) and the range defined by `start_time` and `end_time`. Note that this can return an empty annotation if the ranges don't intersect at all (perhaps raise a warning in this case?)
* `ann.trim()` will have an option `strict` argument, which when set to `True` requires the specified `start_time` and `end_time` to be contained within the range defined by `ann.time` and `ann.duration` (i.e. the trim range must be a sub-range of the annotation's current time range).
* `JAMS.trim()` will result in calling `Annotation.trim()` on all annotations contained in the jam's annotation array. It will *also* update the `jam.file_metadata.duration` field.

In addition, I'm documenting every trim operation in the annotation sandbox (as `ann.sandbox.trim` which stores a list of tuples, each tuple containing the `start_time` and `end_time` of the trim operation) whenever `ann.trim()` is called. If `jams.trim()` is called, the same documentation will occur in `jams.sandbox.trim`, as well as in the sandbox of each annotation stored inside the jam.

**Comment by justinsalamon** [comment]
I just realized something - the current behavior will trim the annotations, but will not modify their start time of the observations. This is useful for when you just want to trim an annotation but not the corresponding audio file. However, if you also trim the audio file, then you'd like the event times to also be adjusted accordingly.

@bmcfee How about we leave the default behavior as this (i.e. observation times are not changed), but add an optional parameter `adjust_times=False` which, if `True`, will also adjust the time of each observation?

**Comment by justinsalamon** [comment]
> How about we leave the default behavior as this (i.e. observation times are not changed), but add an optional parameter adjust_times=False which, if True, will also adjust the time of each observation?

Implemented by 66c1c25, 9f3c4d0, 3dec6c9.

**Comment by justinsalamon** [comment]
Documenting latest modifications to API (as of 95cdda4):
 * trim() has been refactored into trim() and slice()

Here's the low-down for trimming:
 * Input parameters are `start_time`, `end_time`, and `strict=False`
 * `Annotation.trim()` will change the start time/duration of an annotation:
  * new start time is given by `max(Annotation.time, start_time)`
  * new duration is computed using new start time and new end time, where new end time is given by `min(Annotation.time + Annotation.duration, end_time)`
 * `Annotation.trim()` will remove all observations that lie outside of the trim range defined by the new start time and duration:
  * If `strict=False` (default), observations at the boundaries will be adjusted. I.e., if an observation starts before the new start time but ends after it, only the part that lies at/after the new start time is kept. The equivalent happens if the observation starts before but ends after the new end time. 
  * If `strict=True`, observations at the boundaries are also discarded.
 * With the exception of annotations at the boundaries, Annotation.trim() *does not* modify the start time of the observations.
 * Each call to `Annotation.trim()` is documented in `Annotation.sandbox.trim`
 * `JAMS.trim()` calls `AnnotationArray.trim()` which calls `Annotation.trim()` on every annotation in the jams object.
 * `JAMS.trim()` is documented in `JAMS.sandbox.trim`
 * `JAMS.trim()` *does not* affect `JAMS.file_metadata.duration`

Here's the low-down for slicing:
 * Slicing (at the Annotation, AnnotationArray and JAMS level) behaves exactly the same as trimming, except:
  * After trimming, the start time of the annotation is recomputed and is given by `max(0, trimmed_annotation.time - start_time)`, and the start times of all observations it contains are re-computed with respect to this new reference time. To give a concrete example, say an annotation spans time 10-15s. If we call ann.slice(12, 14), the new annotation will have a start time of 0 and duration 2. If instead we call ann.slice(5, 15), the sliced annotation will have a start time of 5 and duration 5.
 * `JAMS.slice()` calls `AnnotationArray.slice()` which calls `Annotation.slice()` on every annotation in the jams object.
  * `JAMS.slice()` *does* affect `JAMS.file_metadata.duration`, which becomes `end_time - start_time`

In a nutshell, trimming is useful when one wishes to modify the time range spanned by an annotation, but keep the corresponding audio file unchanged (and hence all the start times remain unchanged). Slicing is useful when the corresponding audio file will also be trimmed (cropped in time), in which case the start time of the annotation and all observations it contains to need to be adjusted.

**Why doesn't slice() always set the start time of the annotation to 0?**
This could lead to undesirable behavior if you tried to slice an annotation by defining a time range that starts before the start time of the annotation (i.e. `start_time < Annotation.time`). To give a concrete example, say you have an audio file from 0-15s, and an annotation that spans 10-15s. If you called jam.slice(5, 15), the sliced jam file would have a duration of 10s corresponding to "original" time 5-15s. The annotation would have a duration of 5s which corresponds to "original" time 10-15s, but since its start time was reset to 0, it would be as if the annotation describes time 0-5s in the new jam, while it actually describes time 5-10s. That is, it should have a start time of 5s. That's why the new start time of the annotation is given by `max(0, trimmed_annotation.time - start_time)`.
